### ISSUE TITLE
ansible-vault: Dynamic determination of PBKDF2 rounds to speed up encrypt/decrypt

### ISSUE THREAD

**Issue by gebi** [ansible/ansible#20933]
##### ISSUE TYPE
 - Feature Request

##### COMPONENT NAME
ansible-vault

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = <PRIVATE>/src/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
vault_password_file = ./open_vault.sh

##### OS / ENVIRONMENT
N/A

##### SUMMARY
ansible-vault being needlessly slow for decrypt because of PBKDF2 step even when used with high-entropy keys (eg. by key file which is decrypted with gpg through a script outputting it to stdout).

It seems ansible-vault even does one PBKDF2 with 10000 rounds per file encrypt/decrypt also wasting valuable ressoures in an ansible-tower setup.
Where PBKDF2 with 10k rounds would only be necessary ONCE for deriving from a possible weak entropy key (the user vault password) and then working from there. NOT deriving every single password internally for every file again with PBKDF2.

Why is this a problem:
Because most probably ansible is used in a version control system, also most probably one would want to integrate them so showing diff's and merging works with ansible-vault encrypted files.
For this there are quite a few calls to ansible-vault decrypt necessary (and in case of merge also ansible encrypt) where the 10k rounds PBKDF2 makes it needlessly slow!

Even with a running gpg agent ansible-vault wastes nearly a second of time for every ansible-vault file git has to touch. In case of a needed merge there are 3 decrypts and one encrypt necessary increasing the wasted time to 4 seconds per file.

General solution:
Ansible-vault would need a method to set a high entropy key (keyfile) where an PBKDF2 step is unnecessary cutting out the needless latency in case of eg. using ansible-vault with a gpg encrypted keyfile or some other proper password store.

https://github.com/ansible/ansible/blob/devel/lib/ansible/parsing/vault/__init__.py#L680
```python
        b_derivedkey = PBKDF2(b_password, b_salt, dkLen=(2 * keylength) + ivlength,
                            count=10000, prf=pbkdf2_prf)
                                 ^^^^^^^
```

Technical solution:
From a quick view it seems PBKDF2 could be reduced to a much smaller amount of rounds for high entropy keys (single digit number of rounds).
This would possible also lay the foundation for using a key derivation function which resists modern threats against which pbkdf2, with regardlessly of what parameters used, is largely useless. (at the lowest level bcrypt, but the necessary feature would be memory hardness, so either scrypt or argon2).

Fixing this problem would also make #16027 obsolete or at least unnecessary except for very large deployments.

Also mentioning #15060 because it might fit into this bug/discussion.

##### STEPS TO REPRODUCE

for the principle setup: https://blog.erincall.com/p/using-pgp-to-encrypt-the-ansible-vault
for the git integration: https://github.com/building5/ansible-vault-tools

##### EXPECTED RESULTS

It's expected for ansible to NOT being >10 times slower than gpg.

##### ACTUAL RESULTS

```
% time ansible-vault decrypt test.vault.yml --output /dev/stdout
---
# foo
# bar

...
Decryption successful
ansible-vault decrypt test.vault.yml --output /dev/stdout
   0.70s user 0.03s system 98% cpu 0.738 total
```

```
% time gpg --decrypt vault_pass.gpg
SUPER SPECIAL SECRET
gpg --decrypt vault_pass.gpg
   0.05s user 0.00s system 93% cpu 0.056 total
```

**Comment by MarkusTeufelberger** [comment]
See a sample test case in https://github.com/MarkusTeufelberger/ansible/tree/vault-bench (latest 2 commits). Already decrypting merely 20 variables noticeably slows the test execution. There are reports that using vault with dozens or hundreds of files (and probably individually encrypted variables, since they are just embedded files) adds up to several minutes to the deployment. This would be the same problem as in https://github.com/ansible/ansible/issues/8340, only in the general case, where no caching is possible. An example use case would be one vars file or encrypted variable with e.g. a different password per host instead of one global encrypted vars file.

A possible proper solution would be to invent and implement a completely new "Cipher" name (since "AES256" is already in use and fixed to PBKDF2 with 10000 iterations) that has optional rounds/cost parameters. It would probably be wise to wait for https://github.com/ansible/ansible/pull/20566 to land before starting on this, since the cryptography package already offers a reasonably modern hardened KDF (https://cryptography.io/en/latest/hazmat/primitives/key-derivation-functions/#cryptography.hazmat.primitives.kdf.scrypt.Scrypt)

It is by the way not really feasible to dynamically determine the number of pbkdf2 rounds needed for a certain password (also: based on what? Password entropy? CPU resources on the host that happens to create the encrypted file?), especially since this number then needs to be communicated in cleartext to anyone trying to decrypt.

**Comment by MarkusTeufelberger** [comment]
Key derivation functions should be used with high settings if a short human generated password is used or with very low settings, if passwords already contain high entropy (e.g. a reasonably long password file). Currently only PBKDF2 with what I'd call medium settings (too fast and weak to resist a motivated attacker, too slow to provide performance on a single machine) is supported with an individual complete derivation per file.

The proper flow would be to derive high entropy key material from the entered password (using srcypt, PBKDF2 or whatever), then cache this high entropy key and en- or decrypt every file by individually salting this high entropy key material and using that as the local encryption key. The global key material should be salted by a random value that's stored in ansible.cfg or added to every encrypted file/variable, the local salt should be stored per file as it is done already.

Tl;dr:
The core issue: key derivation should happen only once globally per password and resulting files are safe to be encrypted with the same derived key. Currently key derivation happens multiple times for the same password (once per encrypted file/variable) which hurts performance to a point that it is not useful to use more than tens of encrypted files/values in a playbook, while not increasing security.

**Comment by nicoonoclaste** [comment]
Quick question: where would the KDF salt be stored, if the same is to be used for all vault-encrypted files?
In principle, it would be possible to let users set a salt in `ansible.cfg` or somesuch, and have all the performance benefits when it is done, but that's very hack-y.

Aside from that, I would very much be in favour of having
- a new vault “format” (or cipher class, or whatever is the proper name) using more modern crypto;
- parameters such as KDF work factor stored in the encrypted files, and the value used a create/edit/rekey-time taken from `ansible.cfg`.

**Comment by MarkusTeufelberger** [comment]
As I wrote above, either have a "random value" field in ansible.cfg, as many other softwares have or have a format like this for the vault blob: [name of "Cipher"][local settings for cipher, e.g. number of rounds, strength params][local salt][global salt][encrypted data]

If global salt is in ansible.cfg or in every encrypted blob does not make much difference security wise, storage is cheap and it would only be a couple bytes anyways so there is also not much to gain or lose either way.

**Comment by nicoonoclaste** [comment]
@MarkusTeufelberger My point is more that you want the new format to work without breaking backwards compatibility, and you want to store the salt in the files themselves anyhow to support rolling it over.  ;-)

I guess I will just prepare a PR once #20566 lands.

**Comment by samdoran** [comment]
We have no plans to reduce the default number of rounds or make it configurable. Yes, it is slower to decrypt on older hosts, but reducing the number of rounds makes it easier to brute force the ciphertext, reducing overall security.

To discuss this further, please reach out on IRC or the mailing list:

   * IRC: #ansible-devel on irc.freenode.net   
   * mailing list: https://groups.google.com/forum/#!forum/ansible-devel
